src/Modules/OrchardCore.Commerce/Controllers/PaymentController.cs

```diff
@@ -224,57 +223,41 @@ public async Task<IActionResult> SuccessPost(string orderId)
-        }
-        catch (StripeException exception)
-        {
-            return Json(new { error = exception.StripeError.Message });
-        }
-        catch (Exception exception)
-        {
-            _logger.LogError(exception, "Payment processing failed.");
-            var message = T["An error has occurred while processing the payment. Please verify and try again."];
-            return Json(new { error = message.Value });
-        }
+        // Shopping cart ID is null by default currently.
+        await _shoppingCartPersistence.StoreAsync(currentShoppingCart);
 
-        return await GeneratePaymentResponseAsync(paymentIntent);
+        // Set back to default, because a new payment intent should be created on the next checkout.
+        _paymentIntentPersistence.Store(paymentIntentId: string.Empty);
     }
 
-    private async Task<IActionResult> GeneratePaymentResponseAsync(PaymentIntent paymentIntent)
+    [AllowAnonymous]
+    [HttpGet("checkout/middleware")]
+    public async Task<IActionResult> PaymentConfirmationMiddleware([FromQuery(Name = "payment_intent")] string paymentIntent = null)
     {
-        if (paymentIntent.Status == "requires_action" &&
-            paymentIntent.NextAction.Type == "use_stripe_sdk")
+        var fetchedPaymentIntent = await _stripePaymentService.GetPaymentIntentAsync(paymentIntent);
+        var orderId = (await _stripePaymentService.GetOrderPaymentByPaymentIntentIdAsync(paymentIntent))?.OrderId;
+
+        var order = await _contentManager.GetAsync(orderId);
+        var status = order?.As<OrderPart>()?.Status?.Text;
+        var finished = fetchedPaymentIntent.Status == PaymentIntentStatuses.Succeeded &&
+                       order != null &&
+                       status == OrderStatuses.Ordered.HtmlClassify();
+
+        if (finished)
         {
-            // Tell the client to handle the action.
-            return Json(new
-            {
-                requires_action = true,
-                payment_intent_client_secret = paymentIntent.ClientSecret,
-            });
+            await FinalModificationAsync(order);
+
+            return RedirectToAction(nameof(Success), new { orderId });
         }
 
-        if (paymentIntent.Status == "succeeded")
+        if (status == OrderStatuses.PaymentFailed.HtmlClassify())
         {
-            // The payment didn’t need any additional actions and completed!
-            // Create the order content item.
-            var order = await _cardPaymentService.CreateOrderFromShoppingCartAsync(paymentIntent);
-
-            return Json(new { Success = true, OrderContentItemId = order.ContentItemId });
+            return RedirectToAction(nameof(Index));
         }
 
-        // Invalid status.
-        return StatusCode(StatusCodes.Status500InternalServerError, new { error = T["Invalid PaymentIntent status"].Value });
+        return View();
```

Comment:
I think you should only `return View();` if the payment status is `PaymentIntentStatuses.Processing`. You should check that right after the first line of the method to avoid pointlessly querying the order.

Looks like every status besides that is an error if you have reached this point. Including when the payment status is "succeeded" but the `finished` variable above is not true. In this case if you keep looping this site every 2 seconds, then you waste the web server on an impossible request loop, and also waste stripe API requests (I assume there is some quota in production).
![image](https://user-images.githubusercontent.com/4007293/209296224-d7bcb017-5f96-4957-bc47-1d3217ff6313.png)
You should redirect back to the checkout page and display an error notification.

Reply:
Implemented a shortcut here and added error notifier. Although the webhook should work in case the user don't get redirected for any reason, but the payment still goes through.